Enhanced G29

thinkyhead opened the pull request:
- Adapted “Enhanced G29” code referred to in #1499 and posted at
  [3dprintboard.com](http://3dprintboard.com/showthread.php?3105-Auto_Bed_
  Leveling-Enhanced-G29-command)
- Compatible with current G29 while adding some new arguments
  - `V` sets the verbose level for serial out
  - `T` (or `V` > 2) sends a Topology report to serial out
  - `E` works the same way as before
  - `P` works as before (source used `n` or `U`/`u`)
  - `L`, `R`, `B`, `F` work as before
- Added sanity checking for `LRBF`
- Added enum constants for `probe_pt()`'s `retract_action` parameter

thinkyhead committed:
Enhanced G29

- Adapted “Enhanced G29” code referred to in #1499 and posted at
[3dprintboard.com](http://3dprintboard.com/showthread.php?3105-Auto_Bed_
Leveling-Enhanced-G29-command)
- Compatible with current G29 while adding some new arguments
  - `V` sets the verbose level for serial out
  - `T` (or `V` > 2) send a Topology report to serial out
  - `E` works the same way as before
  - `P` works as before (source used `n` or `U`/`u`)
  - `L`, `R`, `B`, `F` work as before
- Still needs sanity checking for `LRBF`

thinkyhead committed:
Arrange ProbeAction values bitwise

thinkyhead commented:
Trying to decide whether to constrain the left, right, back, and front arguments within the bed region and continue with G29, or to throw an error and stop. Probably best to just throw an error.

nophead commented:
Constrain, with a warning in verbose mode. It's amazing how many people
don't realise the probe can't cover the same area as the nozzle despite
having a large offset.

It would be useful if the slicer or host output the region corresponding to
the bounds of the object to get the most accurate levelling, without having
to know the probe offset.

This is how my original levelling on hydraraptor worked. It probes the
active area of the bed clipped to the area the probe can reach, but the
host does all the probing, not the firmware.

On 3 March 2015 at 06:15, Scott Lahteine notifications@github.com wrote:

> Trying to decide whether to constrain the left, right, back, and front
> arguments within the bed region and continue with G29, or to throw an error
> and stop. Probably best to just throw an error.
> 
> ## 
> 
> Reply to this email directly or view it on GitHub
> https://github.com/MarlinFirmware/Marlin/pull/1556#issuecomment-76893038
> .

thinkyhead commented:
Constrain with a warning is nice, because the user can still proceed. But it may confuse users who don't understand why the bed probing is off-center, and who don't see the serial output. That's the only reason I might prefer plain failure. It forces the user to determine that the input is good.

As for hosts managing probing, Marlin would need to provide hooks (G-codes) for manual leveling so the host could handle it. At the very least, the ability to probe at the current point and report back the Z, and the ability to set the bed compensation values in the firmware (which still has to manage the compensation during the print). While it is helpful to provide hooks for host software, it doesn't do much good for those of us who prefer direct SD printing over the use of hosts.

nophead commented:
I was only describing what HydraRaptors host did. For it to work from SD the slicer would have to put the G29 in with the bounding box, but it is better if it doesn't need to know the probe offset, so it gets clipped without error by the firmware, just like speed.

thinkyhead commented:
@nophead Well, I would suggest _not_ putting `G29` into a slicer's preface for generated G-Code, though some users might like to have probing before each print. I expect `G29` mostly to be invoked by a button in a host, or manually by the user from the printer console, just to get initial bed leveling set up. Something Marlin doesn't do, but probably should, is to save the most recent bed leveling results in EEPROM and use them for all subsequent prints.

nophead commented:
If your bed is perfectly flat and never changes then yes you only need to probe once and save it. If it changes with room temperature or humidity then you want to probe per print. If the bed is not perfectly flat you want to probe near the bounds of the object for a more accurate result. It is also faster to probe a smaller area.

I originally developed auto levelling for HydraRaptor, which being made from unsealed wood changed from print to print.

thinkyhead commented:
@nophead That is very RepRap :gem: to accommodate beds made of wood and other porous materials. Good that we make sure to let users make their own best choices without too much limitation.

thinkyhead committed:
Do sanity checking on LRFB values

- Setting boundaries based on configured limits, contrasting with G28
(which forgets these limits?)

thinkyhead commented:
I've added some new code to sanity-check the `L` `R` `F` `B` input values against the hard movement limits.

I see that `M48` is also contributed by the same author. We all love contributors and their contributions, but what is the consensus on including credits like these in the serial output?

```
SERIAL_PROTOCOLPGM("M48 Z-Probe Repeatability test.   Version 2.00\n");
SERIAL_PROTOCOLPGM("Full support at: http://3dprintboard.com/forum.php\n");
```

thinkyhead committed:
Fix wrong defines, broken syntax

thinkyhead committed:
Include 4th extruder support

alexborro commented:
Scott, you read my mind. Yesterday I was testing the M48 on my CoreXY and
had the same thought.

I agree placing the author's name in the source code as comments so
everybody know the responsible.

But in the serial output.. Humm, I don't know.. At most a link to
documentation, not to the main page of a website or forum.

By the way, how is the documentation project? Yesterday I gathered info
about the M48 usage and wanna make it available for all Marlin users.

Cheers.

Alex.
Em 05/03/2015 03:22, "Scott Lahteine" notifications@github.com escreveu:

> I've added some new code to sanity-check the L R F B input values against
> the hard movement limits.
> 
> I see that M48 is also contributed by the same author. We all love
> contributors and their contributions, but what is the consensus on
> including credits like these in the serial output?
> 
> SERIAL_PROTOCOLPGM("M48 Z-Probe Repeatability test.   Version 2.00\n");
> SERIAL_PROTOCOLPGM("Full support at: http://3dprintboard.com/forum.php\n");
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/MarlinFirmware/Marlin/pull/1556#issuecomment-77313330
> .

thinkyhead committed:
Merge branch 'Development' into enhanced_g29
Latest upstream changes

Sniffle commented:
I let roxy know that someone has been working on getting her code merged, she might pop in to answer any questions concerns etc.  Also she has enabled G29 to be saved in eeprom and an eeprom invalidate as well from the same forum location.

thinkyhead commented:
Everything here is now in #1456 so I may just go for it and commit that code. Then I can fix the serial output afterward.

Alex, the documentation project is coming along a bit. I added more details to the Coding Standards document recently, and I've been lurking on the RepRap wiki G-Code page for a while, plotting to go through all the G-Codes and make sure they are up to date. The main `README.md` document needs a refresh on some things, including updates about the latest G-Codes also. It tends to serve as the main location for this documentation, along with the big comment block at the top of `Marlin_main.cpp`. As for our Documentation folder, I haven't visited lately. I wouldn't mind making a comprehensive G-Code document in there with full descriptions of behavior.

thinkyhead commented:
@alexborro I've added [a wiki page for G-Code](https://github.com/MarlinFirmware/Marlin/wiki/Marlin-G-Code) that we can just fill in as we go. For now the documentation style is loose. H3 headers, paragraphs for the main description, and code style for examples. The RepRap wiki automatically makes a nice TOC at the top for header items, but we have to make our own. I've created one as an example.

alexborro commented:
@thinkyhead Thanks for the Wiki, I think it is a good place to document thins.. And once in a while we can generate a PDF file from the wiki and save into Marlin repository.

@Sniffle Tood, ask her to contribute directly to Marlin Repository. It would be nice having someone with her skills helping us improving Marlin.

Cheers.

Alex.

Sniffle commented:
you might have to give her lessons in git, i think she planned on submitting this after the z probe repeatability commit and somehow her fork got all messed up... which doesnt matter now that it is in a new location, but nonetheless i believe that was the issue.  I was actually about to start working with her to get it merged in since i keep a fork up to date in case i want to add anything simple and just happened to notice this today.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -1717,193 +1724,327 @@ void process_commands()
+                  max_probe_x = min(X_MAX_POS, X_MAX_POS + X_PROBE_OFFSET_FROM_EXTRUDER),
+                  min_probe_y = max(Y_MIN_POS, Y_MIN_POS + Y_PROBE_OFFSET_FROM_EXTRUDER),
+                  max_probe_y = min(Y_MAX_POS, Y_MAX_POS + Y_PROBE_OFFSET_FROM_EXTRUDER);
+
```
I think this is the right way to get the limits for probing, because the nozzle isn't allowed to leave the bed area. [The lines around 1670 above](#diff-1cb08de130a6ece2d1b5b9c37bcfef48R1670) seem to be wrong...?

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -2036,6 +2177,7 @@ void process_commands()
         enable_e0();
         enable_e1();
         enable_e2();
+        enable_e3();
```
Somehow these were overlooked in the 4 extruder updates.

alexborro commented:
@Sniffle Git is not a problem.. it looks like strange, but in the end is quite easy.. 
One just need to click Fork button, download the code, work on it, upload to the fork and ask a pull request. 
I can assist her on that, no problem.
My skype: alexborro

Cheers.

Alex.

thinkyhead commented:
I think I will (soon) merge this PR, then do the large G-Code handler patch as a new PR, possibly in smaller pieces (but not one for every g-code!) to make reversions easier.

alexborro commented:
@thinkyhead I'm working in my CoreXY BAL right now.. If you wanna commit it, I can download and test it today.

Cheers.

Alex.

thinkyhead commented:
@alexborro Are you referring to this PR, or to my gcode_handlers PR?

alexborro commented:
@thinkyhead to this PR, as it is related to BAL.

Sniffle commented:
@alexborro I sent her a link to this specific pull request so she could see the comments but i will also send her another PM with your skype info, so that she can contact you if she wishes.

Roxy-3D commented:
@alexborro   

Hi!   The only reason the serial output for the M48 command says 'Full support at 3d-PrintBoard' is because I expected a lot of criticism about it.   First, I was worried that the math types would argue Standard Deviation wasn't being calculated right.   And I was also thinking that people might say "Yeah, having various random legs of travel is OK, but I need some thing else to stress the mechanics."

Mostly, that was in there so I would get a chance to hear what people were thinking and clear up any misconceptions and/or get hints on how to improve the code.

Please just delete that output from the M48 command so everything is consistent with the rest of the code base!

Roxy-3D commented:
Oooops!   I pressed the 'Comment' button too quickly.  I was going to say I don't have a Delta printer.   And almost for sure there are going to be different types of movements that expose repeatability problems in those machines.   I really expected some people to make suggestions on how to move the nozzle for Delta printers but nobody ever said anything.  

So...   If any of you have Delta printers and have suggestions on how to stress the mechanics to expose repeatability issues....    Please chime in!   I'll update the M48 to have a Delta option!   :)

thinkyhead commented:
Thanks for the info @Roxy-3DPrintBoard . Note that the latest code has moved GCode handlers into their own functions. Makes the code a bit neater, but it's a big change to Marlin_main.cpp, so initially any code porting might require a smidgen more work.

redsqrly commented:
Hi, not sure if its just me having a brain freeze, but i believe the G29 E command actually does the exact opposite of what it did in Roxy's original version... For me a major reason for choosing her enhanced command was that the probe stays down for the duration of G29, then z would lift some before the probe is raised after the final probing point. But now even with G29 E the probe is raised right after the final endstop hit causing the probe to scratch the bed messing with repeatability. Is there a setting i missed out?

brainscan commented:
Unfortunately I don't think Roxy's code was copied exactly as she had it, I'm still using my very old version of marlin she made until most of the bugs have gone from the new version. 

Sent from my iPhone

> On 23 Mar 2015, at 17:26, redsqrly notifications@github.com wrote:
> 
> Hi, not sure if its just me having a brain freeze, but i believe the G29 E command actually does the exact opposite of what it did in Roxy's original version... For me a major reason for choosing her enhanced command was that the probe stays down for the duration of G29, then z would lift some before the probe is raised after the final probing point. But now even with G29 E the probe is raised right after the final endstop hit causing the probe to scratch the bed messing with repeatability. Is there a setting i missed out?
> 
> —
> Reply to this email directly or view it on GitHub.

redsqrly commented:
I know, I've been using her old version as well but just thought id give the new one a shot. Everything else seems to work fine except the probe scratch even with E.

thinkyhead commented:
@redsqrly The current code should have that fixed. Probe retraction was moved so that it occurs correctly with `G28` and `G29` in all cases where there is a Z lift set.

redsqrly commented:
@thinkyhead was trying out yesterdays version and was only able to keep the probe down for the duration of G29 with the E parameter, however it still scratches after the last probe as it raises the probe without lifting z. And in Roxy's original enhanced version E parameter was used to engage and disengage the probe at each probing point as per original Marlin auto bed level code. 

"The E parameter: 
OK, first, the Z-Probe isn't going to be engaging and disengaging any more. That is good from an accuracy perspective. But if you want that to happen, you can give the G29 command a parameter E. Both upper and lower case E's will cause engaging and disengaging the probe on each measurement. You can send a G29 E to your printer and it will behave like it used to as it takes measurements."

thinkyhead commented:
I didn't bring over Roxy's G29 verbatim. I just added the new codes. The existing G29 E behavior in Marlin was opposite of Roxy's, as I recall. It's only effect was to keep the probe lowered during probing. That same meaning was retained with the update. I may have that backward, but I'm pretty confident I retained the original meaning. I will compare with Marlin 1.0.2.

redsqrly commented:
Im sorry, this is all pretty new to me and its bit confusing. Im not sure if i understand you correctly... So did Roxy reverse the G29 E behaviour? Ah its all too much, i just want the probe not to scratch :D I don't mind having to type n E to keep the probe down, but the z lift was a pretty big deal for me. Cheers for looking into it and if you find a solution let me know!

Roxy-3D commented:
> Im sorry, this is all pretty new to me and its bit confusing. Im not sure if i understand you correctly... So did Roxy reverse the G29 E behaviour? 

No...  My code to leave the Z_PROBE engaged was there first.   Later, similar code got added to the main Marlin code base but it was reversed.  When I did the original code I figured there was good benefit to leaving the Z_PROBE engaged and decided to make that the default behavior.  But I wanted to allow the original behavior.

When the similar functionality got added to the main Marlin code base, the opposite thinking must have happened.  They wanted to preserve the original behavior but provided the option to over ride that action.

Either approach is fine.  The important fact is the main line code base handles either behavior now!

thinkyhead commented:
@Roxy-3DPrintBoard The word "enhanced" has always seemed too vague. People will wonder, what is being enhanced? The option should have some definite effect, and be documented that way. "E" might stand for Engage, I suppose, but again, vague. There will have to be some discussion about the preferred default behavior.

@redsqrly As for the probe still scratching the bed, that's still a bug. The code that lifts the Z nozzle was moved to `probe_pt()` which is only used for some of the leveling options. I'm looking at making a patch for this shortly.

brainscan commented:
I think Roxy explains it in pretty good detail on the original posts she made at 3dprintboard, couldn't it be copied from there or linked to it? I didn't think any of her work was vague. 

Sent from my iPhone

> On 25 Mar 2015, at 06:31, Scott Lahteine notifications@github.com wrote:
> 
> @Roxy-3DPrintBoard The word "enhanced" has always seemed too vague. People will wonder, what is being enhanced? The option should have some definite effect, and be documented that way. "E" might stand for Engage, I suppose, but again, vague. There will have to be some discussion about the preferred default behavior.
> 
> @redsqrly As for the probe still scratching the bed, that's still a bug. The code that lifts the Z nozzle was moved to probe_pt() which is only used for some of the leveling options. I'm looking at making a patch for this shortly.
> 
> —
> Reply to this email directly or view it on GitHub.

thinkyhead commented:
The word "enhanced" is vague. You will also notice that "E" has the opposite meaning in `M48`, which is also a Roxy contribution. These will need to be squared.

When adding this feature to the main Marlin code-base, I was mindful not to alter the meaning of the "E" argument in `G29` because it would conflict with any online documentation, forums, etc., discussing Marlin. Portions of Roxy's descriptions have made it into the code comments, and should also be included in the documentation page on bed leveling.

Roxy-3D commented:
> @Roxy-3DPrintBoard The word "enhanced" has always seemed too vague. People will wonder, what is being enhanced? The option should have some definite effect, and be documented that way. "E" might stand for Engage, I suppose, but again, vague. There will have to be some discussion about the preferred default behavior.

That thread had multiple 'Enhancements' in it.  It had the ability to leave the probe engaged (by default).  It had the Topographical Map of the bed independent of the user's bed orientation.   And it had the Z_PROBE_LIFT issues resolved.   And it had all of that done just a few months after the 'Accurate Bed Leveling' came out.    At the time, it really was 'Enhanced'.   I guess I could have named the thread 'G29 Changes That Leave Probe Engaged for Better Accuracy With Z_PROBE_LIFT Issues Resolved and a Bonus Topographical Map Thrown Into The Mix'.  But I was short sighted and just called it the 'Enhanced G29'.   And Yes, the 'E' was chosen because the Z_PROBE was going to be E-ngaged for each sampled point.

> The word "enhanced" is vague. You will also notice that "E" has the opposite meaning in M48, which is also a Roxy contribution. These will need to be squared.

Ah...   If you look at just my code, it is all consistent.  The 'E' option causes the Z_PROBE to retract and re-Engage after each point is sampled on both commands.   The reason they 'have opposite' meanings now is not my doing.   If we are going to 'Square Things' I suggest it would make sense to change the current G29 to leave the Z_PROBE engaged by default.  It is difficult to come up with a reason why this is a bad thing, especially when the user has control over it.

thinkyhead commented:
@Roxy-3DPrintBoard I know your code is self-consistent. It just ran counter to the established meaning of `E` in `G29` by the time the feature was merged. I did not presume to reverse this meaning.

redsqrly commented:
Sorry, was working away! @Roxy-3DPrintBoard thanks for the explanation, personally i was excited when i read that your 'G29 Changes That Leave Probe Engaged for Better Accuracy With Z_PROBE_LIFT Issues Resolved and a Bonus Topographical Map Thrown Into The Mix' command was getting merged with marlin, my ABL set up is pretty useless without it and i completely agree that it was a major enchancment. Right now i use your original version to level the bed manually as close as i can get it with with good repeatability and then upload the new version of marlin as i don't have the sd card issues i had with the older version and the servo freeze bug i that i mentioned on 3DPrintBoard had also disappeared with this update...  

@thinkyhead my apologies, but your description of the E command at the top of the thread got me confused, "works the same way as before" makes more sense now. Although right now the meaning of E conflicts with the original forum discussion/online documentation for the enhanced G29 command... Which is obviously a bit confusing. And as for z lift and probe scratch, like i said I'm really new to code and don't really have a clue, so really looking forward to the patch! :) Cheers.

thinkyhead commented:
@redsqrly I'd like to get vote on whether the `E` should have the same meaning that it does now (leave the probe engaged during probing), or if it should be reversed to match `M48` (`engage_probe_for_each_reading`).

NarimaanV commented:
+1 vote to reverse so the probe stays down by default and only engages/disengages when the E parameter is added. Never got why it engaged/disengaged between probes to begin with.

Sniffle commented:
Same +1 to staying engaged
On Mar 27, 2015 10:48 PM, "Narimaan" notifications@github.com wrote:

> +1 vote to reverse so the probe stays down by default and only
> engages/disengages when the E parameter is added. Never got why it
> engaged/disengaged between probes to begin with.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/MarlinFirmware/Marlin/pull/1556#issuecomment-87155797
> .

thinkyhead commented:
I can also add my **+1** to having the probe **stay down by default**.

alvinavr commented:
+1 to stay down by default

redsqrly commented:
+1 on keeping the probe down, and @thinkyhead glad to see you being open to resolving things democratically.

thinkyhead commented:
@redsqrly I also think it's best that UI should follow user expectation. :tophat:

Roxy-3D commented:
> By: ThinkyHead:  @redsqrly I also think it's best that UI should follow user expectation. :tophat:

I think the user interface should be as intuitive as possible.   I think parameters should be able to be more than a single letter!

thinkyhead commented:
@Roxy-3DPrintBoard That would be an interesting extension, at least to clarify it for human readers, but it would add a little to processing time and complexity, and (at least initially) runs afoul of the other firmwares and slicers out there. But, as long as we're at it, we could make all GCode much more readable. Get rid of `G29 E` and replace it with `AutoLevel Extend`. Replace `G28 X` with `Home X`, etc.

GCode is meant to be very concise and primarily machine-generated, plus there is some purism in keeping to a form of GCode that is as familiar to binary load lifters as it is to moisture vaporators. And then, all the firmware projects center around the same GCode specification (I presume, the one on the RepRap wiki), so we'd have to make a new wiki page for it and evangelize the benefits.

Easier, perhaps, is to design a hand-writeable form of GCode, GLang. Some of the intermediate output produced by slicers just before it turns into GCode might look a lot like a GLang. We should after all be able to write GCode in blocks, with variables and loops, just like any proper language, and then we would have something very expressive. Of course, right now we can use any programming language to generate GCode, but making a GCode into its own language would make it a lot more like LOGO — with a graphical view of the GCode as you write it.

Interesting things to consider…

Roxy-3D commented:
> @Roxy-3DPrintBoard That would be an interesting extension, at least to clarify it for human readers, but it would add a little to processing time and complexity, and (at least initially) runs afoul of the other firmwares and slicers out there. 

No...  What I think we should do is add a dozen lines of code.  There would be no extra processing time unless the command is sensitive to and looking for longer parameter names.  I would suggest we prefix all longer parameter names with a ~ so it is clear there is a name there.  If so, we could add the following code block to fully handle in a backward compatible way the long parameter names:

bool code_seen(char *code)  {
  if ( *code !='~' )
      return false;
  strchr_pointer = strstr( cmdbuffer[bufindr], code);
  if ( strchr_pointer==NULL)
      return false;
  strchr_pointer += strlen(code)-1; // Point at the last letter of *code so
                            // code_value() will still work;
  return true;
}

Notice that this alternative defination makes use of the C++ overloading of a function.  Right now every use of code_seen() is looking for a character.  In this case it is looking for a string starting with a ~.   This code never gets called unless a command is sensitive to and using longer parameter names.   To actually use the longer command names the command would do something like:

if (code_seen( "~TEST" ) )  {
    n = code_value();
}

> But, as long as we're at it, we could make all GCode much more readable. Get rid of G29 E and replace it with AutoLevel Extend. Replace G28 X with Home X, etc.

I'm not thinking we 'change' the existing stuff.  100% backward compatibility would be good.  But we could also add a parameter such as G28 ~HOMEX to serve the same purpose as G28 X.   (This is an example!  I don't think this actual example makes sense.  It is for illustration purposes.)  Suddenly the parameters have intuitive meanings and can be used WHERE they make sense.

We had a huge amount of trouble trying to come up with meaningful parameter letters for F-ront, B-ack,L-eft and R-ight on the bounds checking for the G29 extensions.   Having meaningful labels would be a lot easier on people.

> GCode is meant to be very concise and primarily machine-generated, plus there is some purism in keeping to a form of GCode that is as familiar to binary load lifters as it is to moisture vaporators. And then, all the firmware projects center around the same GCode specification (I presume, the one on the RepRap wiki), so we'd have to make a new wiki page for it and evangelize the benefits.

No...  I don't see it that way.   Pretty much, things would continue as they are.  But for more complicated (and probably new) commands there would be the option of having a meaningful name for the specified parameters.    I would expect some parameters have a short and a long form and both are respected by the command.

> Easier, perhaps, is to design a hand-writeable form of GCode, GLang. Some of the intermediate output produced by slicers just before it turns into GCode might look a lot like a GLang. We should after all be able to write GCode in blocks, with variables and loops, just like any proper language, and then we would have something very expressive. Of course, right now we can use any programming language to generate GCode, but making a GCode into its own language would make it a lot more like LOGO — with a graphical view of the GCode as you write it.

No...  I think we maintain 100% backward compatibility.  This would just provide an escape valve for more complicated commands so things can be made clearer.

thinkyhead commented:
Well, the point of a GLang is that it would produce old-school concise GCode, not that it would be executable by an Arduino.
